util/stmtsummary/reader.go

```diff
@@ -0,0 +1,557 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stmtsummary
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pingcap/parser/auth"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/plancodec"
+	"go.uber.org/zap"
+)
+
+// stmtSummaryReader uses to read the statement summaries data and convert to []datum row.
+type stmtSummaryReader struct {
+	user                 *auth.UserIdentity
+	isSuper              bool
+	columns              []*model.ColumnInfo
+	instanceAddr         string
+	ssMap                *stmtSummaryByDigestMap
+	columnValueFactories []columnValueFactory
+}
+
+// NewStmtSummaryReader return a new statement summaries reader.
+func NewStmtSummaryReader(user *auth.UserIdentity, isSuper bool, cols []*model.ColumnInfo, instanceAddr string) *stmtSummaryReader {
+	reader := &stmtSummaryReader{
+		user:         user,
+		isSuper:      isSuper,
+		columns:      cols,
+		instanceAddr: instanceAddr,
+		ssMap:        StmtSummaryByDigestMap,
+	}
+	// initialize column value factories.
+	reader.columnValueFactories = make([]columnValueFactory, len(reader.columns))
+	for i, col := range reader.columns {
+		factory, ok := columnValueFactoryMap[col.Name.O]
+		if !ok {
+			if col.Name.O == util.ClusterTableInstanceColumnName {
+				factory = func(ssElement *stmtSummaryByDigestElement, ssbd *stmtSummaryByDigest) interface{} {
+					return reader.instanceAddr
+				}
```

Comment:
What is this function list used for? a map taking 2 parameters?

Reply:
This is used to set the `INSTANCE` column value, for the cluster_xxx table.